### PR TITLE
feat: add file number and same file path prefix in sftp

### PR DESCRIPTION
**Pull request by Gauravudia**
# Description

- Add file number in file name for each file uploaded to sftp servers per sync.
- Use the same file path prefix for all upload files per sync.
- sample file name 
     - `file_{jobRubID}_{timestamp}_{fileNumber}.csv`
     - `file_cpcnh83xxxxxxxx_1717139665929_2.csv `

## Linear Ticket

Resolves INT-2215
https://linear.app/rudderstack/issue/INT-2215/add-partfilenumber-for-each-sync

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

## Summary by CodeRabbit

- **New Features**
  - Enhanced file upload paths with dynamic placeholders for better organization and traceability.
  - Introduced error handling for file path generation.

- **Bug Fixes**
  - Added validation for file paths to ensure required placeholders are present.

- **Improvements**
  - Updated test cases to handle new metadata fields and error scenarios.
  - Incremented `PartFileNumber` and updated `SourceJobRunID` to improve upload tracking.

- **Code Quality**
  - Introduced new utility functions for better code modularity and readability.

### CODE REVIEW

**Review by koladilip — APPROVED** [review]

**Review by lvrach — COMMENTED** [review]
I don't see how the new behaviour is getting tested

**Comment by Gauravudia** [comment]
> I don't see how the new behaviour is getting tested

Test case are added for validating sample file prefix and file number

**Comment by Gauravudia** [comment]
> I don't see how the new behaviour is getting tested

@lvrach Test case are added for validating sample file prefix and file number

File: router/batchrouter/asyncdestinationmanager/sftp/sftp_test.go [hunk 1]
```diff
@@ -178,20 +180,58 @@ var _ = Describe("SFTP test", func() {
...
+			}
+			defer os.RemoveAll("/tmp/testDir1")
+
+			data := []byte(`{"message":{"action":"insert","fields":{"email":"john@email.com","name":"john"},"messageId":"b6543bc0-f280-4642-8176-4b8d5a1b8fb6","originalTimestamp":"2024-04-20T22:29:37.731+05:30","receivedAt":"2024-04-20T22:29:36.843+05:30","request_ip":"[::1]","rudderId":"853ae90f-0351-424b-973e-a615e6487517","sentAt":"2024-04-20T22:29:37.731+05:30","timestamp":"2024-04-20T22:29:36.842+05:30","type":"record"},"metadata":{"job_id":1}}`)
+			if _, err := tempFile.Write(data); err != nil {
+				log.Fatalf("Failed to write to temporary file: %v", err)
+			}
+
 			received := manager.Upload(&asyncDestination)
+			if err != nil {
```
Comment (lvrach) [inline router/batchrouter/asyncdestinationmanager/sftp/sftp_test.go:231]: error handling should be next to the line where error is returned, is returning and error missed in the line above?

File: router/batchrouter/asyncdestinationmanager/sftp/sftp_test.go [hunk 2]
```diff
@@ -178,20 +180,58 @@ var _ = Describe("SFTP test", func() {
...
+				fmt.Printf("Failed to remove the temporary directory: %v\n", err)
```
Comment (lvrach) [inline router/batchrouter/asyncdestinationmanager/sftp/sftp_test.go:232]: The error handling is inconsistent, some times log.Fatal is a call and others fmt.Printf. 
Given this is a test and gingko is used, we should:
`Expect(err).ShouldNot(HaveOccurred())`

**Review by lvrach — CHANGES_REQUESTED** [review]

File: router/batchrouter/asyncdestinationmanager/sftp/sftp_test.go (hunk 1 above)
Comment (Gauravudia) [reply router/batchrouter/asyncdestinationmanager/sftp/sftp_test.go]: this `err` check is redundant. manager.Upload is not returning the error.

File: router/batchrouter/asyncdestinationmanager/sftp/sftp_test.go (hunk 2 above)
Comment (Gauravudia) [reply router/batchrouter/asyncdestinationmanager/sftp/sftp_test.go]: addressed

**Review by lvrach — APPROVED** [review]

**Pull request merged**